topostats/filters.py

```diff
@@ -208,7 +209,77 @@ def remove_tilt(self, image: np.ndarray, mask: np.ndarray = None):
 
         return image
 
-    def remove_quadratic(self, image: np.ndarray, mask: np.ndarray = None):
+    def remove_nonlinear_polynomial(self, image: np.ndarray, mask: np.ndarray = None) -> np.ndarray:
+        # Script has a lot of locals but I feel this is necessary for readability?
+        # pylint: disable=too-many-locals
+        """Fit and remove a nonlinear polynomial trend of the form a + b * x * y - c * x - d * y
+        from the supplied image.
+
+        Parameters
+        ----------
+        image: np.ndarray
+            2D numpy heightmap array of floats with a polynomial trend to remmove.
+        mask: np.ndarray
+            2D numpy boolean array used to mask out any points in the image that are deemed not to be part of the
+            heightmap's background data. This argument is optional.
+
+        Returns
+        -------
+        np.ndarray
+            Copy of the supplied image with the polynomial trend subtracted.
+        """
+
+        # Define the polynomial function to fit to the image
+        def model_func(x, y, a, b, c, d):
+            return a + b * x * y - c * x - d * y
+
+        image = image.copy()
+        if mask is not None:
+            read_matrix = np.ma.masked_array(image, mask=mask, fill_value=np.nan).filled()
+        else:
+            read_matrix = image
+
+        # Construct a meshgrid of x and y points for fitting to the z heights
+        xdata, ydata = np.meshgrid(np.arange(read_matrix.shape[1]), np.arange(read_matrix.shape[0]))
+        zdata = read_matrix
+
+        # Only use data that is not nan. Nans may be in the image from the
+        # masked array. Curve fitting cannot handle nans.
+        nan_mask = ~np.isnan(zdata)
+        xdata_nans_removed = xdata[nan_mask]
+        ydata_nans_removed = ydata[nan_mask]
+        zdata_nans_removed = zdata[nan_mask]
+
+        # Convert the z data to a 1D array
+        zdata = zdata.ravel()
+        zdata_nans_removed = zdata_nans_removed.ravel()
+
+        # Stack the x, y meshgrid data after converting them to 1D
+        xy_data_stacked = np.vstack((xdata_nans_removed.ravel(), ydata_nans_removed.ravel()))
+
+        # Fit the model to the data
+        # Note: pylint is flagging the tuple unpacking regarding an internal line of scipy.optimize._minpack_py : 910.
+        # This isn't actually an issue though as the extended tuple output is only provided if the 'full_output' flag is
+        # provided as a kwarg in curve_fit.
+        # pylint: disable=unbalanced-tuple-unpacking
```

Comment:
Sorry for not being clearer, need to remove from here as well as adding to the end of the line, otherwise it will be scoped to the rest of the function.

Reply:
Nah you were clear, I just accidentally hit cmd+z before committing 🤦‍♀️